### ISSUE TITLE
About maintaining

### ISSUE THREAD

**Issue by kuraga** [Matt-Esch/virtual-dom#332]
Good day, @Matt-Esch , @Raynos , @neonstalwart !

There are no significant changes in half of year in this repo. I know that you don't have time :smile:  but I see there are many issues, pull requests, ideas, etc.!

You've doing an excellent work! People use `virtual-dom` and we can't stop!!!
Which kind of help is appreciated in this situation? We have to do something! :smile:

**Comment by Zolmeister** [comment]
@Matt-Esch if you need me I'd be glad to help

**Comment by kuraga** [comment]
ping

**Comment by Matt-Esch** [comment]
Yep, I'm super busy these days.

So firstly I believe that feature completeness is a thing and I consider virtual-dom feature complete. A lot of open source modules and languages go through continuous feature additions which are mostly unnecessary. If there is a feature request that can't be implemented outside of virtual-dom then it is worth considering an addition if there is enough support for it. Virtual-dom was deliberately designed without browser-specific features so that user land could build on top of it and so that the scope would be bounded. It serves as a reference implementation for would-be performance pushers.

For some reason there is this opinion in opensource that if you're not actively working on new features then the project is no good. I firmly disagree with this and think it's about time open source focussed on module stability over hapless feature pushing.

That said, I'm going to try to block out some time (in a week or two) to clean up what is left in the issues and then consider what to do about long-term development and maintenance. There's a good chance I will move this project under a different account for long term maintenance, but we shall see.

**Comment by kuraga** [comment]
@Matt-Esch Firstly, great thanks about response!

Secondly I agree with your position.

Thirdly, however, feature completeness doesn't include bug fixes, documentation correction.
And I think, it doesn't include dependencies updating and re-thinking if feature is complete, sometimes.
Let's say that's mantaining...

**Comment by Zolmeister** [comment]
@Matt-Esch There is one feature that I think would have a significant impact, partial subtree diffs, see https://github.com/Matt-Esch/virtual-dom/issues/304#issuecomment-164554945

**Comment by serapath** [comment]
+1 :yum:

**Comment by stevenvachon** [comment]
[inferno](https://npmjs.com/inferno) is a lot faster than this library.

**Comment by kuraga** [comment]
@stevenvachon maybe yes but maybe no.

First, as I can see in benchmark's code, it insert or remove nodes every time and then update. In this case you should use `key` option of virtual-dom to avoid diffing equal but moved nodes. Benchmark doesn't use it. It implicity diff and patch nodes. Inferno (as I see) doesn't diff at all but replaces the whole tree.

Second, in Inferno'es docs we see:
> Unlike other frameworks that use virtual DOMs, Inferno does not "diff" the virtual DOM on each update, but rather it carries out a "diff" on the actual values themselves.

As I understand, Inferno uses optimization available in external [vnode-immutable-thunk](https://github.com/kuraga/vnode-immutable-thunk) (my) package but not in virtual-dom itself.

But yes, this [js-repaint-perfs](http://mathieuancelin.github.io/js-repaint-perfs/) is 2x slower for (my) [RotorJS](https://github.com/kuraga/rotorjs) which uses virtual-dom+vnode-immutable-thunk (but I also didn't use `key` and other optimizations in perf code...)

@Matt-Esch , @Raynos , @neonstalwart, thoughts?

**Comment by kuraga** [comment]
About maintaining: first of all (I think) we have to label each issue (question/feature request/bug) and pull request (bugfix/feature). And then close duplicates...

**Comment by stevenvachon** [comment]
@kuraga check out inferno's ["perf-spike" branch](https://github.com/trueadm/inferno/tree/perf-spike) for the most recent code that they'll soon be releasing.

**Comment by kuraga** [comment]
@stevenvachon I see. I said it is 2x faster than virtual-dom on real app.

According to http://vdom-benchmark.github.io/vdom-benchmark/ Vidom is the fastest.

**Comment by stevenvachon** [comment]
@kuraga yeah, they appeared 50/50 or so to me, but went with inferno because its API involves simple objects instead of functions.

**Comment by kuraga** [comment]
As I investigated, these packages apply changes to DOM during diffing process, while virtual-dom collect vDOM structures during diffing and _then_ re-render the tree applying vDOM. 

Inferno, additionally, produces static content parts which are never diffed. Vidom has some other optimizations.

[[1]](https://github.com/trueadm/inferno/issues/21), [[2]](https://github.com/dfilatov/vidom/issues/171)

**Comment by ghost** [comment]
@Zolmeister  @serapath This project doesn't seem to be maintained. If you are interested in vdom, I'm working on a different project. Drop me an email if you want to join forces!
